src/base/materialized.ts

```diff
@@ -230,60 +236,203 @@ export class MaterializedChatMessage implements IMaterializedNode {
 		return tokenizer.countMessageTokens({ ...this.toChatMessage(), content: '' });
 	});
 
-	private readonly _text = once(() => {
-		let result = '';
+	private readonly _text = once((): (string | MaterializedChatMesageImage)[] => {
+		let result: (string | MaterializedChatMesageImage)[] = [''];
 		for (const { text, isTextSibling } of textChunks(this)) {
+			if (text instanceof MaterializedChatMesageImage) {
+				result.push(text);
+				for (const child of text.children) {
+					if (child instanceof MaterializedChatMessageTextChunk) {
+						result.push(child.text);
+					}
+				}
+				break;
+			}
 			if (
 				text.lineBreakBefore === LineBreakBefore.Always ||
 				(text.lineBreakBefore === LineBreakBefore.IfNotTextSibling && !isTextSibling)
 			) {
-				if (result.length && !result.endsWith('\n')) {
-					result += '\n';
+				let res = result[result.length - 1]
+				if (!(res instanceof MaterializedChatMesageImage) && result.length && !res.endsWith('\n')) {
+					result.push('\n');
 				}
 			}
 
-			result += text.text;
+			if (typeof result[result.length - 1] === 'string') {
+				result[result.length - 1] += text.text;
+			} else {
+				result.push(text.text);
+			}
 		}
 
-		return result.trim();
+		return result;
 	});
 
 	public toChatMessage(): ChatMessage {
+		const content = this.text
+			.filter(element => typeof element === 'string')
+			.join('').trim();
+
+		const images = this._text().filter(element => element instanceof MaterializedChatMesageImage) as MaterializedChatMesageImage[];
+
+		if (images.length > 0) {
+			return {
+				role: ChatRole.User,
+				content: [
+					{ type: 'text', text: content },
+					{
+						type: 'image_url',
+						image_url: { url: getEncodedBase64(images[0].imageUrl), detail: images[0].detail },
+					}],
+			};
+		}
```

Comment:
We should preserve the order of interspersed images and content in the message. This just grabs the first image and puts it at the bottom of the message. You should be able to `this.text.map` and get the right thing.